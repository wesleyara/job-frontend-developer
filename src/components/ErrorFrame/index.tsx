import { MdErrorOutline } from "react-icons/md";

interface ErrorFrameProps {
  error: Error | null;
}

export const ErrorFrame = ({ error }: ErrorFrameProps) => {
  return (
    <section
      data-cy="error-frame"
      className="mt-5 w-full px-3 flex flex-col items-center gap-3"
    >
      <MdErrorOutline size={50} />
      <span className="text-center">{error?.message}</span>
    </section>
  );
};
