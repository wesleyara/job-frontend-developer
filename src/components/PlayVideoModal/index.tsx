import { Dialog, DialogPanel, DialogTitle } from "@headlessui/react";
import { YoutubeDetailsType } from "~/types";
import { Dispatch, SetStateAction } from "react";

interface PlayVideoModalProps {
  item: YoutubeDetailsType;
  isOpen: boolean;
  setIsOpen: Dispatch<SetStateAction<boolean>>;
}

export const PlayVideoModal = ({
  item,
  isOpen,
  setIsOpen,
}: PlayVideoModalProps) => {
  return (
    <Dialog
      open={isOpen}
      onClose={() => setIsOpen(false)}
      className="relative z-50"
    >
      <div className="fixed inset-0 flex mx-auto items-center justify-center p-4">
        <DialogPanel className="space-y-4 border border-gray-300 bg-white p-12 max-w-[768px] w-[90%] text-center rounded-md relative">
          <button
            data-cy="close-youtube-modal"
            className="absolute top-2 right-3"
            onClick={() => setIsOpen(false)}
          >
            X
          </button>
          <DialogTitle className="text-lg">{item.snippet.title}</DialogTitle>

          <iframe
            data-cy="youtube-player"
            src={`https://www.youtube.com/embed/${item.id}`}
            title={item.snippet.title}
            className="inset-0 w-full h-[200px] sm:h-[400px]"
            allowFullScreen
          />
        </DialogPanel>
      </div>
    </Dialog>
  );
};
