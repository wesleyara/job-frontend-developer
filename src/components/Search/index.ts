import { SearchButton } from "./SearchButton";
import { SearchInput } from "./SearchInput";
import { SearchRoot } from "./SearchRoot";

export const Search = {
  Root: SearchRoot,
  Button: SearchButton,
  Input: SearchInput,
};
