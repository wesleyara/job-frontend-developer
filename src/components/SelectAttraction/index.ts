import { SelectAttractionButton } from "./SelectAttractionButton";
import { SelectAttractionContent } from "./SelectAttractionContent";
import { SelectAttractionRoot } from "./SelectAttractionRoot";

export const SelectAttraction = {
  Root: SelectAttractionRoot,
  Content: SelectAttractionContent,
  Button: SelectAttractionButton,
};
