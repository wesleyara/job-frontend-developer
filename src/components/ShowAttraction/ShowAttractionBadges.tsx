import { classificationArray } from "~/lib/utils";
import { Classification } from "~/types";

import { Badge } from "../ui/badge";

interface ShowAttractionBadgesProps {
  classifications: Classification[];
}

export const ShowAttractionBadges = ({
  classifications,
}: ShowAttractionBadgesProps) => {
  return (
    <span className="flex gap-2 justify-center flex-wrap mt-5 px-2">
      {classificationArray(classifications[0]).map(item => (
        <Badge key={item.name} variant="outline">
          {item.name}
        </Badge>
      ))}
    </span>
  );
};
