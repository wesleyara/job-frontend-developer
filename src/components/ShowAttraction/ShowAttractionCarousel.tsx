import { ImagesPerRatioTypes } from "~/types";
import Autoplay from "embla-carousel-autoplay";
import { useWidth } from "essentials-utils";
import Image from "next/image";
import { Dispatch, SetStateAction, useRef } from "react";

import {
  Carousel,
  CarouselContent,
  CarouselItem,
  CarouselNext,
  CarouselPrevious,
} from "../ui/carousel";

interface ShowAttractionCarouselProps {
  ratioSelected: string;
  imagesPerRatio: ImagesPerRatioTypes[];
  selected: ImagesPerRatioTypes | undefined;
  setRatioSelected: Dispatch<SetStateAction<string>>;
}

export const ShowAttractionCarousel = ({
  ratioSelected,
  imagesPerRatio,
  selected,
  setRatioSelected,
}: ShowAttractionCarouselProps) => {
  const plugin = useRef(Autoplay({ delay: 3000, stopOnInteraction: true }));

  const [width, height] = selected?.ratio?.split("_")?.map(Number) || [0, 0];

  const widthSize = useWidth();
  const multiplier = widthSize > 400 ? 80 : 60;
  const ajustedWidth =
    width >= 16 ? width * (multiplier / 4) : width * multiplier;
  const ajustedHeight =
    width >= 16 ? height * (multiplier / 4) : height * multiplier;

  return (
    <span className="flex flex-col items-center gap-4 mt-4 w-full relative">
      <h5>Coleção de imagens</h5>
      <select
        onChange={e => setRatioSelected(e.target.value)}
        value={ratioSelected}
        className="p-2 border border-gray-300 rounded-md bg-white"
      >
        <option value="">Selecione</option>
        {imagesPerRatio.map((item: any) => (
          <option key={item.ratio} value={item.ratio}>
            {item.title}
          </option>
        ))}
      </select>

      {ratioSelected && selected && (
        <span className="flex flex-col items-center">
          <Carousel
            plugins={[plugin.current]}
            className="sm:w-[500px] w-[300px]"
            onMouseEnter={plugin.current.stop}
            onMouseLeave={plugin.current.reset}
          >
            <CarouselContent>
              {selected.images.map(image => (
                <CarouselItem
                  key={image.url}
                  className="flex items-center justify-center"
                >
                  <Image
                    src={image.url}
                    alt={"Carousel image"}
                    width={ajustedWidth}
                    height={ajustedHeight}
                  />
                </CarouselItem>
              ))}
            </CarouselContent>
            {widthSize > 640 && (
              <>
                <CarouselPrevious />
                <CarouselNext />
              </>
            )}
          </Carousel>
        </span>
      )}
    </span>
  );
};
