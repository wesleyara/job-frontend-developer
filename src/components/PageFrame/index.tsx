"use client";

import { useData } from "~/context/DataContext";
import { useState } from "react";
import { FaSearch } from "react-icons/fa";

import { ErrorFrame } from "../ErrorFrame";
import { Search } from "../Search";
import { SelectAttraction } from "../SelectAttraction";
import { ShowAttraction } from "../ShowAttraction";
import { ViewYoutube } from "../ViewYoutube";

export const PageFrame = () => {
  const {
    searchQuery,
    dataTicketSearch,
    dataYoutubeVideosDetails,
    selectedAttraction,
    isLoadingTicketSearch,
    isLoadingYoutubeVideos,
    isLoadingYoutubeVideosDetails,
    errorTicketSearch,
    errorYoutubeVideos,
    errorYoutubeVideosDetails,
    handleChange,
    handleSelectAttraction,
  } = useData();

  const [ratioSelected, setRatioSelected] = useState<string>("");
  const [seeMore, setSeeMore] = useState<boolean>(false);

  const enableToRender = dataTicketSearch && dataTicketSearch.length > 0;

  const render = () => {
    if (isLoadingTicketSearch) {
      return <p className="mt-10">Carregando...</p>;
    }

    if (errorTicketSearch || errorYoutubeVideos || errorYoutubeVideosDetails) {
      return <ErrorFrame error={errorTicketSearch} />;
    }

    if (selectedAttraction) {
      const imagesPerRatio: any = [];

      selectedAttraction.images.forEach(image => {
        const ratio = image.ratio;
        const find = imagesPerRatio.find((item: any) => item.ratio === ratio);

        if (find) {
          find.images.push(image);
        } else {
          imagesPerRatio.push({
            ratio,
            images: [image],
          });
        }
      });

      const selected = imagesPerRatio.find(
        (item: any) => item.ratio === ratioSelected,
      );

      return (
        <>
          <ShowAttraction.Root onClick={() => setSeeMore(prev => !prev)}>
            {!seeMore ? (
              <>
                <ShowAttraction.Info selectedAttraction={selectedAttraction} />
                <ShowAttraction.Badges
                  classifications={selectedAttraction.classifications}
                />
              </>
            ) : (
              <ShowAttraction.Carousel
                setRatioSelected={setRatioSelected}
                selected={selected}
                imagesPerRatio={imagesPerRatio}
                ratioSelected={ratioSelected}
              />
            )}
          </ShowAttraction.Root>

          <span className="flex mt-5">
            {isLoadingYoutubeVideos || isLoadingYoutubeVideosDetails ? (
              <p>Carregando...</p>
            ) : (
              !!dataYoutubeVideosDetails && (
                <ViewYoutube.Root>
                  {dataYoutubeVideosDetails?.map(item => (
                    <ViewYoutube.Card key={item.id} item={item} />
                  ))}
                </ViewYoutube.Root>
              )
            )}
          </span>
        </>
      );
    }

    if (enableToRender) {
      return (
        <div className="flex flex-col gap-3 mt-5 overflow-auto md:w-[600px] w-[90%]">
          {dataTicketSearch.map(item => (
            <SelectAttraction.Root key={item.id}>
              <SelectAttraction.Content item={item} />
              <SelectAttraction.Button
                onClick={() => handleSelectAttraction(item)}
              />
            </SelectAttraction.Root>
          ))}
        </div>
      );
    }
  };

  return (
    <main className="flex flex-col items-center my-24">
      <Search.Root>
        <Search.Input onChange={handleChange} value={searchQuery} />
        <Search.Button icon={FaSearch} />
      </Search.Root>

      {render()}
    </main>
  );
};
