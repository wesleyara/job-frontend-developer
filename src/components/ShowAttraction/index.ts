import { ShowAttractionBadges } from "./ShowAttractionBadges";
import { ShowAttractionInfo } from "./ShowAttractionInfo";
import { ShowAttractionRoot } from "./ShowAttractionRoot";

export const ShowAttraction = {
  Root: ShowAttractionRoot,
  Info: ShowAttractionInfo,
  Badges: ShowAttractionBadges,
};
