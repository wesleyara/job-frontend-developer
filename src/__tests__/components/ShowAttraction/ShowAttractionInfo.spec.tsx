import { render } from "@testing-library/react";
import { attractionMock } from "~/__tests__/mocks";
import { ShowAttraction } from "~/components/ShowAttraction";
import { beforeEach, describe, expect, it } from "vitest";

describe("ShowAttractionInfo unit test", () => {
  let sut: JSX.Element;

  beforeEach(() => {
    sut = <ShowAttraction.Info selectedAttraction={attractionMock} />;
  });

  it("should be defined", () => {
    expect(sut).toBeDefined();
  });

  it("should render correctly", () => {
    const { container } = render(sut);

    expect(container).toMatchSnapshot();
  });

  it("should have a span element", () => {
    const { container } = render(sut);

    expect(container.querySelector("span")).not.toBeNull();
  });

  it("should have a img element", () => {
    const { container } = render(sut);

    expect(container.querySelector("img")).not.toBeNull();
  });

  it("should have a tailwind class name", () => {
    const { container } = render(sut);

    const element = container.querySelector(".flex");

    expect(element).not.toBeNull();
  });

  it("should have a title element", () => {
    const { container } = render(sut);

    expect(container.querySelector("h5")).not.toBeNull();
  });
});
