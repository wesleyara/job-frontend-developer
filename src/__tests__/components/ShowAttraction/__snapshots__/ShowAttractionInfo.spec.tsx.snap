// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ShowAttractionInfo unit test > should render correctly 1`] = `
<div>
  <span
    class="flex items-center gap-4 flex-wrap justify-center"
  >
    <span
      class="w-24 h-24 sm:w-32 sm:h-32 rounded-full relative"
    >
      <img
        alt="Jota Quest"
        class="rounded-full object-cover"
        data-nimg="fill"
        decoding="async"
        fetchpriority="high"
        sizes="100%"
        src="/_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=3840&q=75"
        srcset="/_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=16&q=75 16w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=32&q=75 32w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=48&q=75 48w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=64&q=75 64w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=96&q=75 96w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=128&q=75 128w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=256&q=75 256w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=384&q=75 384w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=640&q=75 640w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=750&q=75 750w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=828&q=75 828w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=1080&q=75 1080w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=1200&q=75 1200w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=1920&q=75 1920w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=2048&q=75 2048w, /_next/image?url=https%3A%2F%2Fs1.ticketm.net%2Fdam%2Fc%2F548%2F5fefbd1c-973b-4b0e-9b2e-d78e4ce37548_106111_TABLET_LANDSCAPE_3_2.jpg&w=3840&q=75 3840w"
        style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
      />
    </span>
    <span
      class="flex flex-col gap-2"
    >
      <h5>
        Jota Quest
      </h5>
    </span>
  </span>
</div>
`;
